server/ERNI.PBA.Server.Host/Services/TransactionCalculator.cs

```diff
@@ -0,0 +1,120 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using ERNI.PBA.Server.DataAccess.Model;
+using ERNI.PBA.Server.Host.Exceptions;
+using Microsoft.AspNetCore.Http;
+
+namespace ERNI.PBA.Server.Host.Services
+{
+    public class TransactionCalculator
+    {
+        private readonly IEnumerable<Budget> _budgets;
+        private readonly int _userId;
+
+        private IList<TeamBudget> _teamBudgets;
+
+        public TransactionCalculator(IEnumerable<Budget> budgets, int userId)
+        {
+            _budgets = budgets;
+            _userId = userId;
+        }
+
+        public IList<Transaction> Transactions { get; private set; }
+
+        public void Calculate(decimal amount)
+        {
+            Calculate(0, amount);
+        }
+
+        public void Calculate(int requestId, decimal amount)
```

Comment:
Still too complicated. It makes sense to separate the calculation.

I just created a simple class for calculation:

`public class RequestAmountCalculator
    {
        public void Calculate<T>(int distributedAmount, IEnumerable<T> list, Func<T, int> getCapacity, Action<T, int> setAmount)
        {
            IEnumerable<T> itemsList = list.OrderBy(_ => getCapacity(_));
            var amountLeft = distributedAmount;

            var count = itemsList.Count();

            while (count > 0)
            {
                var firstItem = itemsList.First();
                var firstItemCapacity = getCapacity(firstItem);

                var amountPerItem = amountLeft / count;

                var amountToDeduct = Math.Min(amountPerItem, firstItemCapacity);

                setAmount(firstItem, amountToDeduct);
                amountLeft -= amountToDeduct;

                itemsList = itemsList.Skip(1);
                count--;
            }
        }
    }`

Reply:
`public class RequestAmountCalculator
    {
        public void Calculate<T>(int distributedAmount, IEnumerable<T> list, Func<T, int> getCapacity, Action<T, int> setAmount)
        {
            IEnumerable<T> itemsList = list.OrderBy(_ => getCapacity(_));
            var amountLeft = distributedAmount;

            var count = itemsList.Count();

            while (count > 0)
            {
                var firstItem = itemsList.First();
                var firstItemCapacity = getCapacity(firstItem);

                var amountPerItem = amountLeft / count;

                var amountToDeduct = Math.Min(amountPerItem, firstItemCapacity);

                setAmount(firstItem, amountToDeduct);
                amountLeft -= amountToDeduct;

                itemsList = itemsList.Skip(1);
                count--;
            }
        }
    }`

Reply:
The Calculate method can distribute the required amount among a set of objects (in our case the budgets).